### PR TITLE
translation class

### PR DESCRIPTION
**Pull request by ltkum**
Concern : Issue #3282 
Based upon the POC from @procrastinatio  in PR #3393 

For now, it is only a translation class. 

The max size of the cache for each language is 10'000, the current number of translations for each language is around 6000. 

I also created a simple method to empty the cache. It might not be used, but I tought it would be best to have one on hand if we want to quickly test if a translation is updated.

To Do List: 
- [x] apply translations to the whole code with this class rather than with the po files and the gettext method
- [x] ensure the cache is thread safe for writing
- [x] cleanup of the code still referring to the old translation method
- [x] cleanup of the translation scripts
- [x] cleanup of the translation files
- [x] cleanup of the makefile
- [x] codebuild (and jenkins) modifications 
Added by momo
- [x] cleanup old tests
- [x] added new tests
- [x] create a test link : [geoadmin test link](https://map.geo.admin.ch/?api_url=%2F%2Fmf-chsdi3.dev.bgdi.ch%2Fltkum_dynamic_translation&lang=fr&topic=ech&bgLayer=ch.swisstopo.pixelkarte-farbe&layers=ch.swisstopo.zeitreihen,ch.bfs.gebaeude_wohnungs_register,ch.bav.haltestellen-oev,ch.swisstopo.swisstlm3d-wanderwege&layers_visibility=false,false,false,false&layers_timestamp=18641231,,,), [api test link](http://mf-chsdi3.dev.bgdi.ch/ltkum_dynamic_translation/rest/services/translations?lang=fr)  
- [x] test this out

### CODE REVIEW

File: chsdi/lib/translator.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+
+class Translator:
+
+    _translations = None
+    _supported_languages = ['en', 'fr', 'de', 'rm', 'it']
+
+    @staticmethod
+    def get_db_connection():
+        return psycopg2.connect(host=os.environ.get('DBHOST'), dbname='bod_master', user='www-data')
```
Comment (ltclm) [inline chsdi/lib/translator.py]: please use the env variable
https://github.com/geoadmin/mf-chsdi3/blob/ltkum_dynamic_translation/rc_dev#L7
for the database staging

Comment (ltkum) [reply chsdi/lib/translator.py]: It has been done.

File: chsdi/lib/translator.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    _translations = None
+    _supported_languages = ['en', 'fr', 'de', 'rm', 'it']
+
+    """
+    Database transation methods
+    """
+    @staticmethod
+    def get_db_connection():
+        return psycopg2.connect(host=os.environ.get('DBHOST'), dbname=os.environ.get('DBSTAGING'), user='www-data')
```
Comment (ltclm) [inline chsdi/lib/translator.py]: thank you but the db name is missing now:
```
'bod_{}'.format(os.environ.get('DBSTAGING'))
```

Comment (ltkum) [reply chsdi/lib/translator.py]: ha yeah, didn't think of that. I correct that

**Comment by procrastinatio** [comment]
So the idea is to get rid of the whole `gettext` business? nice.
But why aren't you using `SQLAlchemy`, but `psychopg2` directly?

**Comment by ltkum** [comment]
Hey @procrastinatio , 
I based off the class from the translation script that was used to create the po files. ( long story short : Something was there and working, I'm using that ^^.) 
I could create a model and use sql alchemy directly.

**Comment by procrastinatio** [comment]
`BeforeRender` and `NewRequest` are used for all sort of black magic. And `add_localizer` is responsible for locale negotiation if I'm not mistaken.

**Comment by ltkum** [comment]
Yeah, I noticed for the black magic ^^. I removed the old translation renderer from subscribers (I FINALLY FOUND IT) and the translation factory. Me might get errors from that, but that would mean cleanup is not completely over

**Comment by procrastinatio** [comment]
@ltkum 
You should have in your `.git/hooks/pre-commit`
```
#!/bin/bash                                                                                                                                                                              
scripts/pre-commit.sh
```

**Comment by procrastinatio** [comment]
You cannot get your connection with env variables.
Put your translation class into something similar to `models.bod.Bod`

**Comment by procrastinatio** [comment]
`lint` finally taming the Martin :-)

**Comment by ltkum** [comment]
> `lint` finally taming the Martin :-)

Do not celebrate victory too soon ^^.

**Comment by procrastinatio** [comment]
@ltkum Just in case, https://github.com/geoadmin/mf-chsdi3/blob/data/chsdi/views/translations.py and https://github.com/geoadmin/mf-chsdi3/blob/data/chsdi/models/bod.py#L491-L501

**Comment by procrastinatio** [comment]
@ltkum @boecklic 

Having a more dynamic translation system is a great thing (because of automatic updates), but I'm still wondering if completely replacing the current system, by a completely new one is the best approach.

The current system is based on `babel` and `gettext`, which are file-based, which is not ideal for our need.
Instead of building something new, I think we may extend the current one and override the [`make_localizer`](https://github.com/Pylons/pyramid/blob/48cf45b0a1ac04e701101fa18778164825edd429/src/pyramid/i18n.py#L169-L174) (`pyramid.i18n.Localizer`), to use translation strings from the DB, either with a caching system or triggered by an API.
Or use the `merge` function of the `Translations` class (same file), to dynamically update the translations with the one from the BOD.

What's you take?

**Comment by ltkum** [comment]
I don't know. I might not want to do an overload approach for one main reason : 
I already did most of the work replacing everything, and I, somehow, don't like to re-do the work.


I am not fond of overloading frameworks functions. It reduce codes readability (If I'm familiar with pyramid and see a "make_localizer" call, I'll expect it to be a standard make localizer call, not an overload.) and make the solution dependent on the used framework. Should we want to switch from pyramid to another framework, we would have another part of the system to redo entirely where a small adjustment might be needed if we separate them from the get go.  

I might be wrong, but I would keep my work as it has been done, and improve upon it.

**Comment by ltkum** [comment]
**The following files have been modified during the merge**
I kept what was on master and used the new translation method where needed. Just to be on the safe side, can those who did the update test out if this still works as intended.

chsdi/templates/htmlpopup/gebietsauslaesse.mako
chsdi/templates/htmlpopup/teileinzugsbiete2.mako
chsdi/templates/htmlpopup/vorfluter.mako

File: chsdi/lib/translator.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    """
+    Database transation methods
+    """
+    @classmethod
+    def get_db_connection(cls):
+        host = cls._host if cls._host is not None else os.environ.get('DBHOST')
+        dbname = "bod_{}".format(cls._staging) if cls._staging is not None else \
+            "bod_{}".format(os.environ.get('DBSTAGING'))
+        return psycopg2.connect(host=host,
```
Comment (ltkum) [inline chsdi/lib/translator.py]: Possible change : using sql alchemy instead of psychopg

File: chsdi/__init__.py [hunk 4]
```diff
@@ -52,7 +53,7 @@ def main(global_config, **settings):
     config.include('pyramid_mako')
 
     # configure 'locale' dir as the translation dir for chsdi app
-    config.add_translation_dirs('chsdi:locale/')
+    # config.add_translation_dirs('chsdi:locale/')
```
Comment (ltkum) [inline chsdi/__init__.py]: TODO : remove these

File: chsdi/lib/translator.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    @classmethod
+    def init(cls, db_host, db_staging, supported_languages):
+        cls._supported_languages = supported_languages.split()
+        cls._host = db_host
+        cls._staging = db_staging
+
+    @classmethod
+    def fill_translations(cls):
+        # using the Lock from threadings to ensure non concurrent writings
+        with cls._lock:
```
Comment (ltkum) [inline chsdi/lib/translator.py]: TODO : error management here.

**Comment by procrastinatio** [comment]
Superseeded by https://github.com/geoadmin/mf-chsdi3/pull/3509

**Pull request closed without merging**